Feature/upgrade hapi

babaalaja opened the pull request:
Bumped  Hapi version to v20. Bumped stale dependencies to latest version and replaced broken packages swagger package to a more maintained [hapi-swagger.](https://github.com/glennjones/hapi-swagger)

To test, complete the following steps
- [ ] yarn 
- [ ]  yarn test
- [ ] yarn bootstrap

Check **/documentation**  route to confirm that swagger docs are displayed.

PR is based off feature/upgrade-mongoose

jabdul commented on lib/server.d.ts:
```diff
@@ -1,4 +1,9 @@
+import Hapi from '@hapi/hapi';
+import { Dict, ServerArgs } from './';
+declare type ServerType = Hapi.Server & {
+  db?: void | typeof import('mongoose');
```
Would we be able to support other dB typings? For example, MySQL or POSTGRESQL.

babaalaja replied:
This is the inferred typing from the call to dbConnect, which is quite strange given that it should return a DbClient. I have done a cast to DbClient. Ran the tests and it worked and should work for other db Typings . Would be pushing a commit in that respect.

Ehbraheem replied:
I think you can reuse the DbType declared somewhere here instead @babaalaja

babaalaja replied:
@Ehbraheem Unfortunately I can't find this type **DbType**.  Is this preferred 
`type ServerType = Hapi.Server & { db?: DbClient; schema?: Dict };
`

Ehbraheem replied:
Yeah. That's what I mean

babaalaja replied:
Ok thanks @Ehbraheem

jabdul commented on package.json:
```diff
@@ -55,20 +52,20 @@
-    "mongoose": "5.9.5",
+    "mongoose": "6.1.8",
     "mysql": "2.15.0",
+    "pino-pretty": "^7.5.1",
```
Could you make update to this file and change the package versions to exact. In fact you may have to fix your '.npmrc' to automatically install exact versions for the future.

babaalaja replied:
Done

jabdul commented on src/bootstrap.ts:
```diff
@@ -37,24 +37,19 @@ const application = (): Promise<Server> =>
     },
-    swaggerUiOptions: {
-      title: 'CRUD API',
-      path: '/docs',
```
Is the default path for swagger UI still `docs`?

babaalaja replied:
/documentation instead.

jabdul commented on src/server.ts:
```diff
@@ -1,17 +1,19 @@
-import { ServerArgs } from './';
+import { DbClient, Dict, ServerArgs } from './';
+
+type ServerType = Hapi.Server & { db?: void | typeof import('mongoose'); schema?: Dict };
```
Same question about other dB types applies here also.

babaalaja replied:
Fixed db now points to DbClient as expected

jabdul commented on src/server.ts:
```diff
@@ -48,16 +49,21 @@ export default async ({
-      { plugin: swaggered, options: swaggerOptions },
-      { plugin: swaggeredUI, options: swaggerUiOptions },
+      { plugin: hapiSwagger, options: swaggerOptions },
+      // { plugin: swaggeredUI, options: swaggerUiOptions },
```
Remove commented out code

babaalaja replied:
Done

jabdul commented on src/users/routes.test.ts:
```diff
@@ -58,8 +59,9 @@ describe(`Routes: ${ROUTE_NAME}`, () => {
-      expect(payload.lastname).toBeDefined();
+      // console.log('The Payload', payload);
+      expect(payload).toBeDefined();
+      // expect(payload.lastname).toBeDefined();
```
Remove commented out code

babaalaja replied:
Done

Ehbraheem commented on src/bootstrap.ts:
```diff
@@ -37,24 +37,19 @@ const application = (): Promise<Server> =>
     },
     swaggerOptions: {
```
It will be better if you have an excellent example of the options that this new swagger package expects.

babaalaja replied:
Again I am not sure on the core expectations here. Does this help [hapi-swagger](https://github.com/glennjones/hapi-swagger/blob/master/usageguide.md)

Ehbraheem replied:
@babaalaja `bootstrap.ts` is a sample app that uses this **@ctt/crud-api** package. What I mean is that you should add the expected options that the `hapi-swagger` package accepts. 

I'm not sure what we have there now still works because it is an option for a different swagger package.

babaalaja replied:
@Ehbraheem It works, the package does not need the accompanying  Swagger UI package that the old Swagger Package required. Did you check **/documentation** when your ran `yarn bootstrap`

Ehbraheem replied:
No, I didn't. Please record a video that shows this 🙏

babaalaja replied:
https://drive.google.com/file/d/1eDDBfZnqvzJGyLcRMh8UJSj7IMhK_Ume/view?usp=sharing

Ehbraheem commented on src/server.ts:
```diff
@@ -79,7 +85,8 @@ export default async ({
 
   app.db = await dbConnect(config);
-  app.schema = schema(app.db);
+  const databaseClient = app.db as unknown;
```
Why the cast here?

babaalaja replied:
dbConnect(config) returns void| import('mongoose') which differs from DbClient which is Promise<Mongoose> | SchemaBuilder. In the new PR I have casted this at the dbConnect layer to DbClient.

Ehbraheem commented on src/users/routes.test.ts:
```diff
@@ -58,8 +59,9 @@ describe(`Routes: ${ROUTE_NAME}`, () => {
 
     it('sets validation on request payload', () => {
       const payload = router.options.validate.payload;
-      expect(payload.firstname).toBeDefined();
```
Please bring back these `.expect` calls; they are crucial. We need to guarantee that this upgrade is not breaking existing functionality.

babaalaja replied:
router.options.validate.payload  changed in  v20. Returns an AST to show the schema definition, hence properties  like firstName are not defined on the root object anymore.

Ehbraheem reviewed (commented):
Good job 👍. Just a few questions and recommendations.

Ehbraheem reviewed (approved):
LGTM 👍. Good job

jabdul reviewed (commented):
@babaalaja Could you review the Readme doc to see if any update to it is required following the upgrades in this PR.

babaalaja commented:
Sure I would look through right away

babaalaja commented:
@jabdul  I have just updated the ReadMe  with updates to the **Swagger docs**, also removed **Halson** dependency in the documentation , couldn't find any reference to it in the project

jabdul commented:
> @jabdul I have just updated the ReadMe with updates to the **Swagger docs**, also removed **Halson** dependency in the documentation , couldn't find any reference to it in the project

Great!

tosin-ojo reviewed (approved):
LGTM
